sdk/resource/resource.go

```diff
@@ -17,30 +17,53 @@
 package resource
 
 import (
-	"reflect"
+	"sort"
+	"strings"
 
 	"go.opentelemetry.io/otel/api/core"
 )
 
 // Resource describes an entity about which identifying information and metadata is exposed.
 type Resource struct {
-	labels map[core.Key]core.Value
+	str    string
+	sorted []core.KeyValue
+	keys   map[core.Key]struct{}
 }
 
 // New creates a resource from a set of attributes.
 // If there are duplicates keys then the first value of the key is preserved.
 func New(kvs ...core.KeyValue) *Resource {
-	res := &Resource{
-		labels: map[core.Key]core.Value{},
-	}
+	res := &Resource{keys: make(map[core.Key]struct{})}
 	for _, kv := range kvs {
-		if _, ok := res.labels[kv.Key]; !ok {
-			res.labels[kv.Key] = kv.Value
+		// First key wins.
+		if _, ok := res.keys[kv.Key]; !ok {
+			res.keys[kv.Key] = struct{}{}
+			res.sorted = append(res.sorted, kv)
 		}
 	}
+	sort.Slice(res.sorted, func(i, j int) bool {
+		return res.sorted[i].Key < res.sorted[j].Key
+	})
+	res.str = buildResourceString(res.sorted)
 	return res
 }
 
+// String implements the Stringer interface and provides a reproducibly
+// hashable representation of a Resource.
+func (r Resource) String() string {
+	return r.str
+}
+
+// Attributes returns a copy of attributes from the resource in a sorted order.
+func (r Resource) Attributes() []core.KeyValue {
+	return append([]core.KeyValue(nil), r.sorted...)
+}
+
+// Equal returns true if other Resource is equal to r.
+func (r Resource) Equal(other Resource) bool {
+	return r.str == other.str
+}
```

Comment:
This can be changed to operate on pointers or not, but would need to delegate to `String()`.

```suggestion
func (r *Resource) Equal(other *Resource) bool {
	return r.String() == other.String()
}
```